// @svgr/cli/lib/util.js
"use strict";

exports.__esModule = true;
exports.transformFilename = transformFilename;
exports.convert = convert;
exports.convertFile = convertFile;
exports.exitError = exitError;
exports.politeWrite = politeWrite;
exports.formatExportName = formatExportName;
exports.CASE = void 0;

var _fs = require("fs");

var _chalk = _interopRequireDefault(require("chalk"));

var _core = _interopRequireDefault(require("@svgr/core"));

var _pluginSvgo = _interopRequireDefault(require("@svgr/plugin-svgo"));

var _pluginJsx = _interopRequireDefault(require("@svgr/plugin-jsx"));

var _pluginPrettier = _interopRequireDefault(require("@svgr/plugin-prettier"));

var _camelcase = _interopRequireDefault(require("camelcase"));

var _dashify = _interopRequireDefault(require("dashify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-console */
const CASE = {
  KEBAB: 'kebab',
  // kebab-case
  CAMEL: 'camel',
  // camelCase
  PASCAL: 'pascal' // PascalCase

};
exports.CASE = CASE;

function transformFilename(filename, filenameCase) {
  switch (filenameCase) {
    case CASE.KEBAB:
      return (0, _dashify.default)(filename.replace(/_/g, '-'), {
        condense: true
      });

    case CASE.CAMEL:
      return (0, _camelcase.default)(filename);

    case CASE.PASCAL:
      return (0, _camelcase.default)(filename, {
        pascalCase: true
      });

    default:
      throw new Error(`Unknown --filename-case ${filenameCase}`);
  }
}

function convert(code, config, state) {
  return _core.default.sync(code, config, { ...state,
    caller: {
      name: '@svgr/cli',
      defaultPlugins: [_pluginSvgo.default, _pluginJsx.default, _pluginPrettier.default]
    }
  });
}

async function convertFile(filePath, config = {}) {
  const code = await _fs.promises.readFile(filePath, 'utf-8');
  return convert(code, config, {
    filePath
  });
}

function exitError(error) {
  console.error(_chalk.default.red(error));
  process.exit(1);
}

function politeWrite(program, data) {
  if (!program.silent) {
    process.stdout.write(data);
  }
}

function formatExportName(name) {
  if (/[-]/g.test(name) && /^\d/.test(name)) {
    return `Svg${(0, _camelcase.default)(name, {
      pascalCase: true
    })}`;
  }

  if (/^\d/.test(name)) {
    return `Svg${name}`;
  }

  if (/[-]/g.test(name)) {
    return (0, _camelcase.default)(name, {
      pascalCase: true
    });
  }

  return name;
}
